File: 242. Valid Anagram/Solution.java
public class Solution {

  /**
   * 
   * @param s
   * @param t
   * @return
   *         Time Complexity: O(nlogn), n is the length of s
   *         Space Complexity: O(1)
   */
  public boolean isAnagramSorting(String s, String t) {
    if (s.length() != t.length()) {
      return false;
    }
    char[] str1 = s.toCharArray();
    char[] str2 = t.toCharArray();
    Arrays.sort(str1);
    Arrays.sort(str2);
    return Arrays.equals(str1, str2);
  }

  /**
   *
   * @param s
   * @param t
   * @return
   *         Time Complexity: O(n), n is the length of s
   *         Space Complexity: O(1)
   */
  public boolean isAnagramFreqCounter(String s, String t) {
    if (s.length() != t.length()) {
      return false;
    }
    int[] table = new int[26];
    for (int i = 0; i < s.length(); i++) {
      table[s.charAt(i) - 'a']++;
    }
    for (int i = 0; i < t.length(); i++) {
      table[t.charAt(i) - 'a']--;
      if (table[t.charAt(i) - 'a'] < 0) {
        return false;
      }
    }
    return true;
  }
}

File: 242. Valid Anagram/solution.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 * Time Complexity: O(nlogn), n is the length of the strings s and t
 * Space Complexity: O(1)
 */
var isAnagramSorting = function(s, t) {
    if (s.length !== t.length) {
        return false;
    };
    const str1 = s.split('').sort().join('');
    const str2 = t.split('').sort().join('');

    return str1 === str2;
};

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagramFreqCounter = function(s, t) {
    if (s.length !== t.length) {
        return false;
    };
    
    const freqCounter = {};
    for (let char of s) {
        freqCounter[char] = (freqCounter[char] || 0) + 1;
    }
    for (let char of t) {
        if (!freqCounter[char]) {
            return false;
        }
        freqCounter[char]--;
    }
    return true;
};